introduction_to_amazon_algorithms/jumpstart-foundation-models/text-generation-benchmarking/benchmarking/load_test.py

```diff
@@ -0,0 +1,344 @@
+        return statistics
+
+    @staticmethod
+    def _collect_statistics(data: List[Union[float, int]]) -> Dict[str, Any]:
+        return {
+            "Median": np.median(data).item(),
+            "Average": np.average(data).item(),
+            "Minimum": np.amin(data).item(),
+            "Maximum": np.amax(data).item(),
+            "p50": np.quantile(data, 0.50).item(),
+            "p90": np.quantile(data, 0.90).item(),
+            "p95": np.quantile(data, 0.95).item(),
+        }
+
+
+class LoadTester:
+    def __init__(
+        self,
+        predictor: Predictor,
+        payload: Dict[str, Any],
+        model_id: str,
+        payload_name: str,
+        tokenizer_model_id: Optional[str] = None,
+        price_per_endpoint: Optional[float] = None,
+    ) -> None:
+        self.predictor = predictor
+        self.payload = payload
+        self.model_id = model_id
+        self.payload_name = payload_name
+        self._logging_prefix = logging_prefix(self.model_id, self.payload_name)
+        if tokenizer_model_id is not None:
+            self.tokenizer = AutoTokenizer.from_pretrained(tokenizer_model_id)
+        else:
+            self.tokenizer = None
+        self.price_per_endpoint = price_per_endpoint
+
+    def predict_once_and_collect_client_results(self) -> PredictionResult:
+        """Perform a single endpoint prediction and produce a PredictionResult."""
+        time_utc_start = datetime.datetime.utcnow()
+        try:
+            result = self.predictor.predict(self.payload, custom_attributes="accept_eula=True")
+        except Exception as e:
+            result = e
+        time_utc_end = datetime.datetime.utcnow()
+        return PredictionResult(time_utc_start, time_utc_end, self.payload, result)
+
+    def run_load_test(self, num_invocations: int, max_workers: int) -> Optional[BatchInvocationStatistics]:
+        """Concurrently invoke an endpoint prediction multiple times and gather results in BatchInvocationStatistics."""
+        time_utc_start = datetime.datetime.utcnow()
+        results = []
+        error = None
+        timeout_seconds = SM_INVOCATION_TIMEOUT * num_invocations / max_workers
+        with futures.ThreadPoolExecutor(max_workers=max_workers) as executor:
+            futures_list = [
+                executor.submit(self.predict_once_and_collect_client_results) for _ in range(num_invocations)
+            ]
+            for future in futures.as_completed(futures_list, timeout=timeout_seconds):
+                result = future.result()
+                if isinstance(result.result, Exception):
+                    error = copy.deepcopy(result.result)
+                    for future_to_cancel in futures_list:
+                        future_to_cancel.cancel()
+                else:
+                    results.append(result)
+
+        if error is not None:
+            raise error
```

Comment:
are we not allowing endpoints to return transient errors during the load test

Reply:
Right now, if any invocation raises an error, the whole load test fails. These errors typically fall in 3 categories: CUDA issues, SageMaker 60s invocation timeout, and parameter validation checks. Each of these reasons would justify ending the concurrency probe. Do you have other transient errors in mind?